plutus-core/cost-model/CostModelGeneration.md

```diff
@@ -59,48 +62,380 @@ Generating a cost model for CPU time involves a number of steps.
+experimentation may be required to achieve this, and it may not always be
+possible to satisfy both goals simultaneously.  In such cases it may be
+necessary to sacrifice some accuracy in order to guarantee security.
+
+
+#### Step 6: add code to read the costing function from R into Haskell 
+
+Next we have to update the code which converts benchmarking results into JSON
+models.  Go to
+[`CreateBuiltinCostModel`](./create-cost-model/CreateBuiltinCostModel.hs) and add
+entries for the new builtin in builtinCostModelNames
+
+```
+  , paramXorByteString                   = "xorByteStringModel"
+```
+(Getting the string wrong here, for example putting "xorByteString" instead will
+give `parse error (not enough input) at ""`. Errors will occur whenever the
+Haskell code attempts to read something from an R object that doesn't actually
+occur in the object, and they can sometimes be quite cryptic.)
+
+Also add a new clause in [`CreateBuiltinCostModel`](./create-cost-model/CreateBuiltinCostModel.hs):
+
+```
+    paramXorByteString                   <- getParams xorByteString paramXorByteString
+```
+
+and a function to extract the cost parameters for the R code.  This should be modelled on the existing
+functions at the end of the file:
+
+```
+xorByteString :: MonadR m => (SomeSEXP (Region m)) -> m (CostingFun ModelTwoArguments)
+xorByteString cpuModelR = do
+  cpuModel <- ModelTwoArgumentsMinSize <$> readModelMinSize cpuModelR
+  let memModel = ModelTwoArgumentsMinSize $ ModelMinSize 0 1
+  pure $ CostingFun (cpuModel) memModel
+```
+
+The CPU costing function is obtained by running the R code, but the memory usage
+costing function is defined statically here.  Memory usage costing functions
+only account for memory retained after the function has returned and not for any
+working memory that may be allocated during its execution.  Typically this means
+that the memory costing function should measure the size of the object returned
+by the builtin.  For our `xorByteString` implementation, if the arguments have
+sizes `m` and `n` then the result will have size `min(m,n)` so we define the memory
+costing function to be `(m,n) -> 0 + 1*min(m,n)`.
+
+
+#### Step 7: update the R code
+
+We now have to extend the R code in [`models.R`](./data/models.R).  Firstly, add
+an entry for the arity of the builtin in the `arity` function:
+
+```
+   arity <- function(name) {
+       switch (name,
+           "AddInteger" = 2,
+           ...
+           "XorByteString" = 2,
+           ...
+           )
+```
+
+Now add a function to infer coefficients for the CPU costing function from
+benchmarking data.  In the case of `xorByteString` we assume that the time taken
+will be linear in the minimum of the sizes of the arguments (ie, the arguments
+of the new builtin).  It is often worthwhile to plot the benchmark data and
+experiment with it in order to check that in has the form expected when the
```

Comment:
that in has -> that it has

Reply:
Endless typos!  Thanks for spotting that.